[GStreamer] Buffering hysteresis and buffering improvements for Broadcom

eocanha opened the pull request:
#### aededeb7ddd9e55b53d9eda0be721ab3fd8180e4
<pre>
[GStreamer] Buffering hysteresis and buffering improvements for Broadcom
<a href="https://bugs.webkit.org/show_bug.cgi?id=275683">https://bugs.webkit.org/show_bug.cgi?id=275683</a>

Reviewed by Xabier Rodriguez-Calvar.

The usage of a fixed buffering level to evaluate the need for buffering
results in a too unstable behaviour, with quick changes between states
that may create stuttering.

This commit adds hysteresis to the buffering level, with low and high
watermark levels that trigger &quot;buffering needed&quot; (when below the low
watermark), &quot;buffering completed&quot; (when above high watermark) and &quot;same
state as before&quot; when the level is between both marks.

In addition to that, the pipeline is automatically paused on a seek when
working on stream mode (using GstQueue2 for partial in-memory buffering
instead of using GstDownload for full on-disk download of the media).
This is because, when seeking, rebuffering must occur from scratch to
get the data for the new target position, and buffering level starts
from 0% in that case, which means that buffering is needed and the
pipeline must be paused. Both the m_isBuffering flag and
m_bufferingPercentage are immediately forced to true and 0%, respectively.
The previous values (see explanation below) are also set to the same
values by using the new resetHistory parameter in updateBufferingStatus().
This makes sure that the &quot;end of buffering&quot; condition can be properly
detected.

The buffering status and buffering level computation has been
centralized in updateBufferingStatus(). Both the status and the level
now store the last value, so it&apos;s possible to check if there&apos;s been any
difference on them and to decide here on the hysteresis status
evolution.

updateStates() is where the consequences of buffering changes happen.
This method now relies in the status and level updates explained in the
previous paragraph. In the specific case of an async change, the current
values for m_isBuffering and m_bufferingPercentage are ignored (previous
ones are used). This is to be able to detect significant changes in the
hysteresis (watermark level crossing) after the async change has
completed. Otherwise, we would be detecting those crossings in a moment
where nothing can be done (pipeline state changes are ignored can can&apos;t
be enforced during async state changes).

Some extra improvements have been added to adapt to the special
circumstances of Broadcom devices, where the PlayPump Nexus component
can store a lot of data which is hidden to the vanilla buffering
algorithm. Now that data is had into account to compute the realistic
buffering level. Also, more agile low/high watermarks of 20% and 80% are
used for buffering in the case of playbin2 in stream buffering mode on
Broadcom.

20-80% levels for low/high watermark are now used on all devices
(Broadcom and not Broadcom), since usage tests proved that this makes
the buffering more agile also on Raspberry Pi, for instance.

All these changes are compatible with the Quirks framework. A new
needsBufferingPercentageCorrection() method has been added to enable the
extra buffering logic (which corrects the queue2 buffering level with the
buffer levels of PlayPump and multiqueue) on the platforms that use Nexus.
All the code to implement that logic has been hidden in the
GStreamerQuirkBroadcomBase class, from which both GStreamerQuirkBcmNexus
and GStreamerQuirkBroadcom inherit.

That code needs some elements that in principle would be very coupled with
MediaPlayerPrivateGStreamer, such as vidfilter, queue2, multiqueue and a
MovingAverage to smooth buffering levels. To avoid that coupling, they
are stored in an opaque GStreamerQuirkState, stored by the player private
but implemented by a GStreamerQuirkBroadcomBaseState subclass that only
GStreamerQuirkBroadcomBase knows how to interpret (by casting after
ensuring that the casting is legal).

There&apos;s also the problem that multiple competing quirks can be installing
at the same time, but all this buffering percentage correction code must
only be run by one (only one!) of them. To allow the theoretical coexistence
of multiple quirk state, each quirk instance has one associated state, and
the player private maintains that correspondence using a HashMap indexed by
quirk. To ensure that only one quirk is run at each time, the family of methods
in GStreamerQuirksManager only forward the call to the first configured
GStreamerQuirk subclass that declares the need to apply corrections. Any other
configured subclass is ignored.

Finally, some drive-by improvements are present in this commit. For instance,
converting some const char* to ASCIILiteral here and there.

* Source/WebCore/platform/SourcesGStreamer.txt: Added GStreamerQuirkBroadcomBase.
* Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.h:
(WebCore::gstStructureGetArray): Added.
* Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
(WebCore::gstStructureGetArray): Method to get a GstArray GstValue stored into a GstStructure. Returns a Vector of the specified type, which includes basic types and &quot;const GstStructure*&quot;.
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::play): Make code compatible with InitiallyPaused.
(WebCore::MediaPlayerPrivateGStreamer::doSeek): Pause pipeline and reset buffering level to 0% in case we&apos;re operating in stream mode.
(WebCore::MediaPlayerPrivateGStreamer::queryBufferingPercentage): Get the buffering percentage from the most relevant element and delegate to the quirks.
(WebCore::MediaPlayerPrivateGStreamer::fillTimerFired): Clarify that this method is only used in on-disk buffering. Rely on queryBufferingPercentage() now, instead of manually getting the value from here.
(WebCore::MediaPlayerPrivateGStreamer::handleMessage): Delegate to the quirks to store/delete references to specific elements in the quirks state.
(WebCore::MediaPlayerPrivateGStreamer::processBufferingStats): Delegate to the quierks to apply buffering percentage corrections on platforms that need them.
(WebCore::MediaPlayerPrivateGStreamer::updateBufferingStatus): Update the values of m_wasBuffering, m_isBuffering, m_previousBufferingPercentage, m_bufferingPercentage and m_didDownloadFinish. The first 4 attributes should *ONLY* be modified from this method, in order to guarantee coherency. The only *EXCEPTION* to this rule is the delay which can happen in updateStates(). The new resetHistory parameter sets the value both in the previous value family of attributes and in the current value family of variables. Also, restart the fillTimer if download has restarted (it was never restarted before this patch!)
(WebCore::MediaPlayerPrivateGStreamer::updateStates): Assume that m_wasBuffering, etc. have been precomputed by updateBufferingStatus() before updateStates() is called. In the case of an async state change, delay m_isBuffering and m_bufferingPercentage &quot;to the past&quot; (m_wasBuffering and m_previousBufferingPercentage), so that any change on them can be detected on the next iteration.
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.h: Added new m_wasBuffering and m_previousBufferingPercentage attributes to store previous values of m_isBuffering and m_bufferingPercentage and ease detection of changes. Added quirks state.
(WebCore::MediaPlayerPrivateGStreamer::shouldDownload): Needed by quirks.
(WebCore::MediaPlayerPrivateGStreamer::setQuirkState): Configure a GStreamerQuirkState for a specific GStreamerQuirk on the player private (each quirk has its own state, and in theory there can be many quirk-state stored).
(WebCore::MediaPlayerPrivateGStreamer::quirkState): Get the quirks state associated to a specific quirk from the player private, or nullptr if there&apos;s no configured state for that quirk.
* Source/WebCore/platform/gstreamer/GStreamerHolePunchQuirkBcmNexus.h: identifier() is now const ASCIILiteral.
* Source/WebCore/platform/gstreamer/GStreamerHolePunchQuirkFake.h: Ditto.
* Source/WebCore/platform/gstreamer/GStreamerHolePunchQuirkRialto.h: Ditto.
* Source/WebCore/platform/gstreamer/GStreamerHolePunchQuirkWesteros.h: Ditto.
* Source/WebCore/platform/gstreamer/GStreamerQuirkAmLogic.h: Ditto.
* Source/WebCore/platform/gstreamer/GStreamerQuirkBcmNexus.h: Ditto. Inherit from base class.
* Source/WebCore/platform/gstreamer/GStreamerQuirkBroadcom.h: Ditto. Inherit from base class.
* Source/WebCore/platform/gstreamer/GStreamerQuirkBroadcomBase.cpp: Added. Base class for GStreamerQuirkBcmNexus and GStreamerQuirkBroadcom.
(WebCore::GStreamerQuirkBroadcomBase::GStreamerQuirkBroadcomBase): Constructor.
(WebCore::GStreamerQuirkBroadcomBase::queryBufferingPercentage const): Get the buffering percentage from Queue2 if present and the player shouldn&apos;t download (streaming mode). Returns the name of the element finally used to query the percentage to, or nullptr if none was queried.
(WebCore::GStreamerQuirkBroadcomBase::correctBufferingPercentage const): On systems using PlayPump (eg: Broadcom Nexus), correct the buffering percentage with the data stored in multiqueue and PlayPump in order to make the buffering percentage more realistic. Uses the MovingAverage. Queries vidfilter and audfilter if needed, so it also works in audio or video-only pipelines.
(WebCore::GStreamerQuirkBroadcomBase::resetBufferingPercentage const): Sets the percentage to a given value in the MovingAverage.
(WebCore::GStreamerQuirkBroadcomBase::setupBufferingPercentageCorrection const): Gets/deletes references to vidfilter, multiqueue and queue2 in the quirks state.
(WebCore::GStreamerQuirkBroadcomBase::ensureState const): Retrieves GStreamerQuirkState associated to this quirk, creating one and storing it on MediaPlayerPrivateGStreamer if not already present.
* Source/WebCore/platform/gstreamer/GStreamerQuirkBroadcomBase.h: Added.
(WebCore::GStreamerQuirkBroadcomBase::needsBufferingPercentageCorrection const): True for this class and derived ones.
(WebCore::GStreamerQuirkBroadcomBase::MovingAverage::MovingAverage): Initialize an array of &quot;length&quot; number of elements, asserting if length is too big that the values would create overflow when computing the average.
(WebCore::GStreamerQuirkBroadcomBase::MovingAverage::reset): Reset all the values to a given value (usually zero).
(WebCore::GStreamerQuirkBroadcomBase::MovingAverage::accumulate): Insert a new value in the array and recompute the moving average.
* Source/WebCore/platform/gstreamer/GStreamerQuirkRealtek.h: identifier() is now const ASCIILiteral.
* Source/WebCore/platform/gstreamer/GStreamerQuirkRialto.h: Ditto.
* Source/WebCore/platform/gstreamer/GStreamerQuirkWesteros.h: Ditto.
* Source/WebCore/platform/gstreamer/GStreamerQuirks.cpp: Added new methods.
(WebCore::GStreamerQuirksManager::GStreamerQuirksManager): Adapt to ASCIILiteral quirk identifier.
(WebCore::GStreamerQuirksManager::createAudioSink): Ditto.
(WebCore::GStreamerQuirksManager::createWebAudioSink): Ditto.
(WebCore::GStreamerQuirksManager::createHolePunchVideoSink): Ditto.
(WebCore::GStreamerQuirksManager::isHardwareAccelerated const): Ditto.
(WebCore::GStreamerQuirksManager::audioVideoDecoderFactoryListType const): Ditto.
(WebCore::GStreamerQuirksManager::getAdditionalPlaybinFlags const): Ditto.
(WebCore::GStreamerQuirksManager::needsBufferingPercentageCorrection const): True if any configured quirk needs it. The first configured quirk that needs it will be the one responsible to fulfill all the requests for this family of methods. The other ones will be ignored, even if they also need percentage correction. This is to avoid more than one implementation doing the corrections, as we don&apos;t want them to accumulate and over-correct.
(WebCore::GStreamerQuirksManager::queryBufferingPercentage const): Forward query request to the first configured quirk needing correction.
(WebCore::GStreamerQuirksManager::correctBufferingPercentage const): Forward correction request to the first configured quirk needing correction.
(WebCore::GStreamerQuirksManager::resetBufferingPercentage const): Ditto, but for reset.
(WebCore::GStreamerQuirksManager::setupBufferingPercentageCorrection const): Ditto, but for setup.
* Source/WebCore/platform/gstreamer/GStreamerQuirks.h:
(WebCore::GStreamerQuirkBase::GStreamerQuirkState::GStreamerQuirkState): Constructor.
(WebCore::GStreamerQuirk::needsBufferingPercentageCorrection const): Default empty implementation.
(WebCore::GStreamerQuirk::queryBufferingPercentage const): Ditto.
(WebCore::GStreamerQuirk::correctBufferingPercentage const): Ditto.
(WebCore::GStreamerQuirk::resetBufferingPercentage const): Ditto.
(WebCore::GStreamerQuirk::setupBufferingPercentageCorrection const): Ditto.
* LayoutTests/platform/glib/TestExpectations: Reported new failing test as <a href="https://bugs.webkit.org/show_bug.cgi?id=280069">https://bugs.webkit.org/show_bug.cgi?id=280069</a>, but IMHO it&apos;s not a regression. The test was passing before by sheer luck.

Canonical link: <a href="https://commits.webkit.org/284072@main">https://commits.webkit.org/284072@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2c94695b47d393e07a0ef8bae68fb226c24a6dfb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/68417 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/47809 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/21076 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/72484 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/19562 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/70534 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/55605 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/19378 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/54628 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/13031 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/71484 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/43710 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/59080 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/35091 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/40378 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/16500 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/17919 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/62336 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/16848 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/74176 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/12388 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/16114 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/62079 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/12427 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/59158 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/62100 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15166 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/10022 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/3637 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/43610 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/44684 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/45878 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/44426 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

calvaris commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.h:
```diff
@@ -522,6 +522,44 @@ class MediaPlayerPrivateGStreamer
+            int sum = 0;
+            for (unsigned char i = 1; i < m_length; i++) {
+                m_values[i-1] = m_values[i];
+                sum += m_values[i-1];
```
`[i - 1]`

calvaris commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.h:
```diff
@@ -522,6 +522,44 @@ class MediaPlayerPrivateGStreamer
+            return sum / m_length;
+        }
+    private:
+        const unsigned char m_length;
```
If you want a small unsigned int, please use `uint8_t`, everywhere. chars are misleading

calvaris commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.h:
```diff
@@ -655,6 +697,10 @@ class MediaPlayerPrivateGStreamer
+    GRefPtr<GstElement> m_vidfilter;
+    GRefPtr<GstElement> m_multiqueue;
+    GRefPtr<GstElement> m_queue2;
+    MovingAverage m_streamBufferingLevelMovingAverage = MovingAverage(10);
```
You should be able to do:

`MovingAverage m_streamBufferingLevelMovingAverage(10);`

nikolaszimmermann replied:
s/ = Moving../ { 10 }/

calvaris commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.h:
```diff
@@ -522,6 +522,44 @@ class MediaPlayerPrivateGStreamer
+        }
+    private:
+        const unsigned char m_length;
+        int* m_values;
```
please, use a Vector. You can do the initial allocation for it in the constructor.

calvaris commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -1313,6 +1327,116 @@ void MediaPlayerPrivateGStreamer::commitLoad()
+    if (!originalBufferingPercentage && m_queue2) {
+        guint currentLevelBytes = 0;
+        g_object_get(m_queue2.get(), "current-level-bytes", &currentLevelBytes, nullptr);
+        correctedBufferingPercentage1 = currentLevelBytes > maxSizeBytes ? 100 : (int)(currentLevelBytes * 100 / maxSizeBytes);
```
Please use c++ casts.

calvaris commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -1313,6 +1327,116 @@ void MediaPlayerPrivateGStreamer::commitLoad()
+        GUniqueOutPtr<GstStructure> stats;
+        g_object_get(m_multiqueue.get(), "stats", &stats.outPtr(), nullptr);
+        const GValue* queues = gst_structure_get_value(stats.get(), "queues");
+        guint size = gst_value_array_get_size(queues);
```
unsigned or even auto.

calvaris commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -1313,6 +1327,116 @@ void MediaPlayerPrivateGStreamer::commitLoad()
+        g_object_get(m_multiqueue.get(), "stats", &stats.outPtr(), nullptr);
+        const GValue* queues = gst_structure_get_value(stats.get(), "queues");
+        guint size = gst_value_array_get_size(queues);
+        for (guint i = 0; i < size; i++) {
```
ditto

calvaris commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -1313,6 +1327,116 @@ void MediaPlayerPrivateGStreamer::commitLoad()
+
+    // Current-level-bytes seems to be inacurate, so we compute its value from the buffering percentage.
+    size_t currentLevelBytes = (size_t)maxSizeBytes * (size_t)originalBufferingPercentage / (size_t)100
+        + (size_t)playpumpBufferedBytes + (size_t)multiqueueBufferedBytes;
```
c++ casts

calvaris commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -1313,6 +1327,116 @@ void MediaPlayerPrivateGStreamer::commitLoad()
+    // Current-level-bytes seems to be inacurate, so we compute its value from the buffering percentage.
+    size_t currentLevelBytes = (size_t)maxSizeBytes * (size_t)originalBufferingPercentage / (size_t)100
+        + (size_t)playpumpBufferedBytes + (size_t)multiqueueBufferedBytes;
+    correctedBufferingPercentage2 = currentLevelBytes > maxSizeBytes ? 100 : (int)(currentLevelBytes * 100 / maxSizeBytes);
```
ditto

calvaris commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -1313,6 +1327,116 @@ void MediaPlayerPrivateGStreamer::commitLoad()
+{
+    GRefPtr<GstQuery> query = adoptGRef(gst_query_new_buffering(GST_FORMAT_PERCENT));
+
+    bool queryOk = false;
```
isQueryOk

calvaris commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -2023,6 +2144,58 @@ void MediaPlayerPrivateGStreamer::handleMessage(GstMessage* message)
+
+                // Also get the multiqueue (if there's one) attached to the vidfilter. We'll need it later to correct the buffering level.
+                GstPad* sinkPad = nullptr;
+                GstIterator* iter = gst_element_iterate_sink_pads(m_vidfilter.get());
```
GUniquePtr and GstIteratorAdaptor

calvaris commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -2199,7 +2372,15 @@ void MediaPlayerPrivateGStreamer::processBufferingStats(GstMessage* message)
+            percentage = correctBufferingPercentage(percentage);
+    }
+
+    updateBufferingStatus(mode, double(percentage));
```
c++ casts

nikolaszimmermann commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.h:
```diff
@@ -522,6 +522,44 @@ class MediaPlayerPrivateGStreamer
+
+        ~MovingAverage()
+        {
+            free(m_values);
```
s/free/delete[]/ :-)

calvaris reviewed (changes requested):
This is a general comment for the whole patch. This pattern has to disappear:

`if (quirksManager.isEnabled() && quirksManager.needsPlaypumpBufferingLogic())`

What you need to do is creating specific functions in the quirks to perform the actions you need and then ask the quirks if they need to do anything additional to what is done. The whole purpose of the quirks is hiding platform quirks in the quirks files. Here you are not hiding.

I am aware that broadcom and bcmnexus might have common code that which these changes broadcom and bcmnexus could have a quite big part of common code so I think to overcome this, you might need to create a BCMBase class to put the common code and make both quirks inherit from it.

calvaris commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -1313,6 +1328,50 @@ void MediaPlayerPrivateGStreamer::commitLoad()
+    GST_TRACE_OBJECT(pipeline(), "[Buffering] %s reports %d buffering", elementName, percentage);
+
+    ASSERT(mode == GST_BUFFERING_DOWNLOAD);
+    if (mode != GST_BUFFERING_DOWNLOAD)
```
If you're checking the mode, you don't need to ASSERT in advance, I would just do an `ASSERT_NOT_REACHED();` inside the if after the `GST_WARNING`

calvaris commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -1313,6 +1328,50 @@ void MediaPlayerPrivateGStreamer::commitLoad()
     updateStates();
 }
 
+bool MediaPlayerPrivateGStreamer::queryBufferingPercentage(GstBufferingMode& mode, int &percentage)
```
Please, change this to return `std::optional<int>` .

And do you actually want to allow changing the buffering mode inside this function? I see below you are ASSERTing and after that you are not checking it back again.

eocanha replied:
No, I don't want. And I don't actually need any previous percentage value passed as a parameter. It was just a "tool" to use a variable defined in the caller scope (variable passed by reference) that I don't actually need, because the percentage is always overwritten. I'm removing the parameters.

calvaris commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -2212,44 +2275,71 @@ void MediaPlayerPrivateGStreamer::updateMaxTimeLoaded(double percentage)
+void MediaPlayerPrivateGStreamer::updateBufferingStatus(GstBufferingMode mode, double percentage, bool resetHistory)
 {
-    bool wasBuffering = m_isBuffering;
+    // m_wasBuffering, m_isBuffering, m_previousBufferingPercentage and m_bufferingPercentage can ONLY be modified from this method.
```
Then write this in the .h file where this attributes are declared.

calvaris commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.h:
```diff
@@ -566,8 +578,12 @@ class MediaPlayerPrivateGStreamer
     RefPtr<TextureMapperPlatformLayerProxy> m_platformLayerProxy;
 #endif
 #endif
+    // Can only be changed from updateBufferingStatus().
```
Clarify this as I asked above.

calvaris commented on Source/WebCore/platform/gstreamer/GStreamerHolePunchQuirkFake.h:
```diff
@@ -29,7 +29,7 @@ namespace WebCore {
 class GStreamerHolePunchQuirkFake final : public GStreamerHolePunchQuirk {
 public:
-    const char* identifier() final { return "FakeHolePunch"; }
+    const char* identifier() const final { return "FakeHolePunch"; }
```
Nice drive-by improvement!

eocanha replied:
Thank the compiler, which started to pester me with const correctness messages in the other const methods where I attempted to call this one.

calvaris replied:
Nice! Let's go a bit further (yes, you're hating me right now and I accept it) and make this return `ASCIILiteral`.

calvaris commented on Source/WebCore/platform/gstreamer/GStreamerQuirkBcmNexus.h:
```diff
@@ -22,18 +22,20 @@
     std::optional<bool> isHardwareAccelerated(GstElementFactory*) final;
     std::optional<GstElementFactoryListType> audioVideoDecoderFactoryListType() const final { return GST_ELEMENT_FACTORY_TYPE_PARSER; }
     Vector<String> disallowedWebAudioDecoders() const final { return m_disallowedWebAudioDecoders; }
+    bool needsBufferingPercentageCorrection() const { return true; }
```
If I am not mistaken, this is already in the base class, right?

calvaris commented on Source/WebCore/platform/gstreamer/GStreamerQuirkBroadcomBase.cpp:
```diff
@@ -0,0 +1,192 @@
+    GST_DEBUG_CATEGORY_INIT(webkit_broadcom_base_quirks_debug, "webkitquirksbroadcombase", 0, "WebKit Broadcom Base Quirks");
+}
+
+bool GStreamerQuirkBroadcomBase::queryBufferingPercentage(MediaPlayerPrivateGStreamer* playerPrivate, const char*& elementName, GRefPtr<GstQuery>& query) const
```
This looks like it's time for std::optional<const char*> or even just const char* with nullptr for false

eocanha replied:
Maybe, but it would make the logic in https://github.com/WebKit/WebKit/pull/29998/commits/0aae941a74fbc7be84ce9ddffa73b7779bff0388#diff-4376b854280205494dc9de0f08f985afd4d6584a7487c0a136adac2954c003cdR1340 more complex. That elementName is a variable that comes from the caller site and has its own interpretation and usage rules, as well as query.

calvaris replied:
Can't you do a `value_or` at the caller?

eocanha replied:
Ended up using `const char*` and nullptr to signal "query not successful".

calvaris commented on Source/WebCore/platform/gstreamer/GStreamerQuirkBroadcomBase.cpp:
```diff
@@ -0,0 +1,192 @@
+        GUniqueOutPtr<GstStructure> stats;
+        g_object_get(state.m_multiqueue.get(), "stats", &stats.outPtr(), nullptr);
+        const GValue* queues = gst_structure_get_value(stats.get(), "queues");
+        unsigned size = gst_value_array_get_size(queues);
```
Please, define gstStructureGet<usize> and use it.

calvaris replied:
sorry, there is already unsigned, you don't need to define it, so you can use it already.

eocanha replied:
In this very specific case (and only in this) gstStructureGet() doesn't make sense. That function is useful to avoid defining an extra variable to be passed to gst_structure_get_whatevertype() by reference, but in this case we're getting a GValue that holds a GArray. Those aren't supported by gstStructureGet(). Adding support for generic `const GValue*` values in gstStructureGet() doesn't make sense, because those are already returned by gst_structure_get_value() as the result of the function call (or nullptr if not present). They don't need to be passed as arguments by reference.

Therefore, I would just be changing one call by another to get the same type in the end.

However, using gstStructureGet() in the lines below to get basic types **does** make sense.

calvaris commented on Source/WebCore/platform/gstreamer/GStreamerQuirkBroadcomBase.cpp:
```diff
@@ -0,0 +1,192 @@
+        unsigned size = gst_value_array_get_size(queues);
+        for (unsigned i = 0; i < size; i++) {
+            unsigned bytes = 0;
+            if (gst_structure_get_uint(gst_value_get_structure(gst_value_array_get_value(queues, i)), "bytes", &bytes))
```
you can use gstStructureGet here as well with `value_or(0)` and it gets easier to read, IMHO

calvaris commented on Source/WebCore/platform/gstreamer/GStreamerQuirkBroadcomBase.cpp:
```diff
@@ -0,0 +1,192 @@
+    // This code must support being run from different GStreamerQuirkBroadcomBase subclasses without breaking. Only the
+    // first subclass instance should run.
+    if (currentState == GST_STATE_NULL && newState == GST_STATE_READY
+        && g_strstr_len(GST_ELEMENT_NAME(element), 13, "brcmvidfilter")) {
```
We should be using here `gst_element_get_name`, put it into a GUniquePtr and wrap it in a StringView to check if it `startsWith`. If you don't think checking the name is going to have threading issues, you can still wrap it into a StringView.

But maybe, given what I saw on e52172278726d91ff43890425d8821a73a332975 , I think we should be better checking the type instead of the name.

And I see more appearances of this below, so please check all your uses of `g_strstr_len` to apply the new pattern.

eocanha replied:
Ended up checking for class name.

calvaris commented on Source/WebCore/platform/gstreamer/GStreamerQuirkBroadcomBase.cpp:
```diff
@@ -0,0 +1,192 @@
+                GRefPtr<GstElement> peerElement = adoptGRef(GST_ELEMENT(gst_pad_get_parent(peerSrcPad.get())));
+                // The multiqueue reference is useless if we can't access its stats (on older GStreamer versions).
+                if (peerElement && g_strstr_len(GST_ELEMENT_NAME(peerElement.get()), 10, "multiqueue")
+                    && g_object_class_find_property(G_OBJECT_GET_CLASS(peerElement.get()), "stats"))
```
`gstObjectHasProperty`

calvaris commented on Source/WebCore/platform/gstreamer/GStreamerQuirkBroadcomBase.cpp:
```diff
@@ -0,0 +1,192 @@
+    }
+}
+
+bool GStreamerQuirkBroadcomBase::isEnsuredOwnedState(MediaPlayerPrivateGStreamer* playerPrivate) const
```
Maybe I am misunderstanding the whole point of the structure is:

1. quirks are singletons
2. we need to store quirk specific information for each media player instance, which is a challenge if they are singletons.

Your implementation serves the purpose but I think we can do better:

1. We have the GStreamerQuirkState class.
2. We keep a `HashMap<GStreamerQuirk*, std::unique_ptr<GStreamerQuirkState>>` inside the media player.
3. When a quirk needs to keep state, it pushes one to the hash map via set/get methods in the player.

eocanha replied:
I understand what you're asking for and I also thought about that approach (one state per quirk, multiple quirks, all of them held by the player in a map) **but**, in this specific case of actions performed by the GStreamerQuirkBroadcomBase superclass of GStreamerQuirkBroadcom and GStreamerQuirkBcmNexus, that approach poses a problem:

All these buffering percentage corrections are "accumulative"/"destructive". They take the original percentage, ask some data to some extra pipeline elements, update a moving average and return the "decorated" percentage. In a system that needs both the GStreamerQuirkBroadcom and GStreamerQuirkBcmNexus quirks, letting the two of them "correct" the percentage would mean that the original percentage would be corrected (increased/decreased) twice, and that's wrong. Both quirks are competing for the same correction, and only **one** of them must do it (no matter which one as long as one does it). That, on systems that have only the BcmNexus quirk, on systems that have only the Broadcom quirk, and on systems that have both.

The solution that I could come with, stores a single state, belonging to a single quirk (the first one who attempts to configure the state on the player) and disallows the other competing quirks to operate on that state. This ensures that at most one quirk performs the correction.

I agree that storing multiple states, one per quirk, would be interesting in the future for those use cases, just not for this specific one.

calvaris replied:
From what I understand, this correction is done in the base class so it does not matter if it's performed by any of the instantiated subclasses. Then, when running the quirk, you should perform just one correction and bail out when that correction is done. I understand you are already doing that at `GStreamerQuirksManager::correctBufferingPercentage` as you perform the correction and then return.

If that were not enough because you might correct the percertage and still get the same result, which would force the second correction, then that method signature should be changed to maybe return an `std::optional<int>` for example and the moment you get a valid value, you bail out.

With all this in mind, I think you can keep my suggestion without convoluting the structures in such a way you proposed to fix this small inconvenience. Am I wrong?

eocanha replied:
You are right about not correcting twice by using an optional<int>, but I'm not so happy about all the subclasses setting up the state and having to keep references to several elements. If only one subclass should actually operate on each player, why wasting space keeping a state for the idle subclass (idle as in "lost the competition to be in charge of correcting the percentage")?

Maybe I can come up with a way to support multiple states, but prevent the last subclass that arrives to the party against configuring itself and operating. It just feels like a lot of overengineering for a problem that should actually be simpler.

calvaris replied:
Maybe you can do something, yes, but I think we are trying to over-engineer here to save just an idle structure. A compromise here could be trying the change the signature of the hash map to take an ASCIILiteral as key and then the base class could install the structure, which could be the same for both subclasses. That way we install the structures we need when we need them.

eocanha replied:
<del>Ok. As we discussed in person, I'm going to change all the methods in GStreamerQuirk related to buffering percentage correction to tell in some way if they have processed the request or not. That way, the loops in GStreamerQuirksManager that ask all the registered quirks can bail out as soon as one of the quirks attends the request (returns true or returns a value on it s std::optional). This will effectively prevent more than one GStreamerQuirkBroadcomBase subclass instance to be in charge of buffering level correction, as intended.</del>

In the end, I realized about the right place to enforce that only one subclass takes care of correcting the buffering percentage: in [the loops of the methods of GstQuirksManager](https://github.com/WebKit/WebKit/pull/29998/files/0aae941a74fbc7be84ce9ddffa73b7779bff0388#diff-6cff3097ce7ab86a5680c2bd50e386e51592786422838ad844891f56fa5c9cdbR288). The first configured quirk that reports that needsBufferingPercentageCorrection() will be the one that always takes care of the processing. As the configured quirks list is expected not to change, the same subclass instance will be the one which takes care (the first in the m_quirks vector). This removes the need of arbitrary bool/std::optional return values where they aren't strictly needed.

calvaris commented on Source/WebCore/platform/gstreamer/GStreamerQuirkBroadcomBase.h:
```diff
@@ -0,0 +1,96 @@
+            m_values[m_values.size() - 1] = value;
+            sum += value;
+            intmax_t result = sum / m_values.size();
+            ASSERT(result < std::numeric_limits<int>::max());
```
I think this overflow check is wrong. If I am not mistaken, if you add two numbers and it overflows, what you are going to get is a negative number so always smaller than max.

Anyway, from what I understand, you are adding natural numbers so first, you can use uintmax_t. Then to check the overflow, you just check if the sum of the two numbers is smaller than one of them, then you know you overflew.

eocanha replied:
No, no. Have into account that I'm using intmax_t for the operations (which I hope is much larger than the regular int type and won't ever overflow), but then I want to fit that result in the regular int type. I perform the check in "the intmax_t world".

Now let's consider some actual numbers: The default MovingAverage size used in this patch is 10. That means a maximum of 10 * 100 = 1000 accumulated value. To make this computation overflow in a 64 bit signed int, we would need a MovingAverage with 9223372036854775807 / 100 = 92233720368547758 items. No computer can hold that array in memory in the foreseable future (famous last words, when computers had 640KB of RAM).

Maybe I can add a constructor-time assert to ensure that the caller isn't ever trying to construct a MovingAverage with more than, let's say, 50 elements, and also ensure that values can't be greater than 100. If I do that, I could even avoid the need of intmax_t.

calvaris commented on Source/WebCore/platform/gstreamer/GStreamerQuirks.h:
```diff
@@ -44,7 +46,30 @@ class GStreamerQuirkBase {
+    private:
+        // For identification purposes only, to avoid the wrong GStreamerQuirkBase subclass to use a state that
+        // doesn't belong to itself. The pointer shouldn't be used to access the instance.
+        const void* m_owner = nullptr;
```
If you use what I suggested above, this is no longer needed.

eocanha replied:
Yes, but see my argument about why allowing a single state per player makes sense for this specific use case.

eocanha replied:
In the end, I used the HashMap that you suggested and this isn't needed.

philn commented on Source/WebCore/platform/gstreamer/GStreamerQuirks.h:
```diff
@@ -44,7 +46,30 @@ class GStreamerQuirkBase {
     virtual ~GStreamerQuirkBase() = default;
 
-    virtual const char* identifier() = 0;
+    virtual const char* identifier() const = 0;
```
Maybe in a follow-up PR this could be made to return an `ASCIILiteral`?

calvaris replied:
Why wait to a follow up? :stuck_out_tongue_winking_eye:

eocanha commented on Source/WebCore/platform/gstreamer/GStreamerQuirkBroadcomBase.cpp:
```diff
@@ -0,0 +1,192 @@
+
+    // The Nexus playpump buffers a lot of data. Let's add it as if it had been buffered by the GstQueue2
+    // (only when using in-memory buffering), so we get more realistic percentages.
+    if (mode != GST_BUFFERING_STREAM || !state.m_vidfilter)
```
I'm going to review myself here to avoid forgetting about [a problem detected downstream](https://github.com/WebPlatformForEmbedded/WPEWebKit/issues/1309#issuecomment-2223461375):

> The Broadcom buffer percentage correction isn't applied in the case of Deezer because that is an audio-only service, and the code looks for vidfilter (not present). I can work on trying to apply the same corrections when only audfilter is present.

So: Maintain references to m_audfilter, check if any of m_vidfilter or m_audfilter is present, and ask for buffered-bytes to the first of them. Also, the property is declared in the [original source code](https://github.com/Metrological/bcm-collaboration/blob/master/reference/vidfilter/src/gst_brcm_vid_filter.c#L452) as "buffered_bytes" (with underscore), instead of "buffered-bytes" (with minus sign). I would swear it worked anyway, but I have to double-check.

calvaris commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.h:
```diff
@@ -249,6 +249,14 @@ class MediaPlayerPrivateGStreamer
 
     String codecForStreamId(const String& streamId);
+    bool shouldDownload() { return m_fillTimer.isActive(); }
+    void setQuirkState(const GStreamerQuirk* owner, GStreamerQuirkBase::GStreamerQuirkState&& state) { m_quirkStates.set(owner, WTF::makeUnique<GStreamerQuirkBase::GStreamerQuirkState>(WTFMove(state))); }
```
What's the problem of moving the structure inside the map without the pointer? You have to change the declaration and maybe declaring some more contructors/operators in the class.

eocanha replied:
The problem is that the HashMap documentation doesn't explain how to get a reference to the stored element, instead of the element itself (what does "the element itself" actually means? a copy of it? WTFMove() is nice to store things, but not to retrieve them. I don't want a copy, I want the real element (or a reference, or a pointer) that I can change and the changes stay inside the instance stored in the HashMap).

Can you point me to the documentation that explains how that can be done, or to a commit that has done that before?

That's why I thought that storing a pointer (specifically, an auto-destructible one, that is, unique_ptr) made more sense in this case, where I need to share GStreamerQuirkState between multiple clients.

calvaris replied:
You're right. Forget about this.

calvaris commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.h:
```diff
@@ -249,6 +249,14 @@ class MediaPlayerPrivateGStreamer
+    void setQuirkState(const GStreamerQuirk* owner, GStreamerQuirkBase::GStreamerQuirkState&& state) { m_quirkStates.set(owner, WTF::makeUnique<GStreamerQuirkBase::GStreamerQuirkState>(WTFMove(state))); }
+    GStreamerQuirkBase::GStreamerQuirkState* quirkState(const GStreamerQuirk* owner)
+    {
+        if (!m_quirkStates.contains(owner))
+            return nullptr;
+        return m_quirkStates.get(owner);
```
You can use getOptional() here and if you manage to make the structure in place inside the map, you can just return in the function `std::optional<GStreamerQuirkState&>`.

eocanha replied:
No, I can't. [Optionals can't hold references](https://en.cppreference.com/w/cpp/utility/optional#:~:text=There%20are%20no%20optional%20references,nullopt_t%20or%20std%3A%3Ain_place_t.).

calvaris commented on Source/WebCore/platform/gstreamer/GStreamerQuirkBroadcomBase.cpp:
```diff
@@ -0,0 +1,200 @@
+    GST_DEBUG_CATEGORY_INIT(webkit_broadcom_base_quirks_debug, "webkitquirksbroadcombase", 0, "WebKit Broadcom Base Quirks");
+}
+
+const char* GStreamerQuirkBroadcomBase::queryBufferingPercentage(MediaPlayerPrivateGStreamer* playerPrivate, GRefPtr<GstQuery>& query) const
```
Can't this be a const GRefPtr& .  I don't think you need the non-const &

calvaris commented on Source/WebCore/platform/gstreamer/GStreamerQuirkBroadcomBase.cpp:
```diff
@@ -0,0 +1,200 @@
+            multiqueueBufferedBytes += gstStructureGet<unsigned>(
+                gst_value_get_structure(gst_value_array_get_value(queues, i)),
+                "bytes"_s).value_or(0);
+        }
```
I think it would be interesting that you implement a `Vector<T> gstStructureGetArray<T>(GstStructure*)` at GstCommon and apply it here. gstStructureGet adapters are a trend these days.

calvaris commented on Source/WebCore/platform/gstreamer/GStreamerQuirkBroadcomBase.cpp:
```diff
@@ -0,0 +1,200 @@
+    state.m_streamBufferingLevelMovingAverage.reset(bufferingPercentage);
+}
+
+void GStreamerQuirkBroadcomBase::setupBufferingPercentageCorrection(MediaPlayerPrivateGStreamer* playerPrivate, GstState currentState, GstState newState, GstElement* element) const
```
I am wondering if it could be a good idea to make the `GstElement*` into a `GRefPtr<GstElement>&&` and then move when necessary.

calvaris commented on Source/WebCore/platform/gstreamer/GStreamerQuirkBroadcomBase.cpp:
```diff
@@ -0,0 +1,200 @@
+            // We'll need it later to correct the buffering level.
+            for (auto* sinkPad : GstIteratorAdaptor<GstPad>(GUniquePtr<GstIterator>(gst_element_iterate_sink_pads(element)))) {
+                GRefPtr<GstPad> peerSrcPad = adoptGRef(gst_pad_get_peer(sinkPad));
+                if (peerSrcPad) {
```
if (!peerSrcPad)
    continue;

eocanha replied:
In this specific case, where the loop only has one possible iteration, `continue` would work as intended (the loop would exit because there aren't more pending iterations).

calvaris commented on Source/WebCore/platform/gstreamer/GStreamerQuirkBroadcomBase.cpp:
```diff
@@ -0,0 +1,200 @@
+                                // Now we hopefully have peerElement pointing to the multiqueue.
+                                peerElement = adoptGRef(GST_ELEMENT(gst_pad_get_parent(peerSrcPad.get())));
+                            }
+                            break;
```
Isn't this loop going to iterate just once with an unconditional break at the end?

eocanha replied:
Yes, that's the expected thing to happen. Both GstBrcmVidFilter and GstBrcmAudFilter have a single sink pad, so the loop would have only one iteration. But don't focus on that. Focus on how beautiful the ":" syntax is to get the pad. 🥺

calvaris commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.h:
```diff
@@ -655,6 +678,7 @@ class MediaPlayerPrivateGStreamer
     RefPtr<PlatformMediaResourceLoader> m_loader;
 
     RefPtr<GStreamerQuirksManager> m_quirksManagerForTesting;
+    HashMap<const GStreamerQuirk*, std::unique_ptr<GStreamerQuirkBase::GStreamerQuirkState>> m_quirkStates;
```
I see a couple of problems with this approach:

1. The Broadcom base class will instantiate two structures, no big deal.
2. If we want to specialize the state for the broadcom subclasses, boom.

Anyway, I can live with these problems and solve them later should we need it.

eocanha replied:
No problem. I'm now declaring all the private stuff in GStreamerQuirkBroadcomBase as protected (that's needed to have all the GStreamerQuirkBroadcomBaseState dependencies visible to subclasses) and ensureState() is now virtual. Any subclass can now override ensureState(), with the condition that it creates a (hypothetical) subclass of GStreamerQuirkBroadcomBaseState, sets it on the player private, and returns it cast as GStreamerQuirkBroadcomBaseState.

Anyway, even if the code now solves that problem, it's still a theoretical one, since all the percentage correction family of functions in GStreamerQuirksManager is designed to forward the calls to one (but never more than one) subclass. There should be no clashes between subclasses.

calvaris commented on Source/WebCore/platform/gstreamer/GStreamerQuirkBroadcomBase.h:
```diff
@@ -0,0 +1,95 @@
+    GStreamerQuirkBroadcomBase();
+
+    bool needsBufferingPercentageCorrection() const { return true; }
+    const char* queryBufferingPercentage(MediaPlayerPrivateGStreamer*, GRefPtr<GstQuery>&) const;
```
ASCIILiteral

calvaris commented on Source/WebCore/platform/gstreamer/GStreamerQuirks.cpp:
```diff
@@ -274,6 +274,61 @@ bool GStreamerQuirksManager::shouldParseIncomingLibWebRTCBitStream() const
+    return false;
+}
+
+const char* GStreamerQuirksManager::queryBufferingPercentage(MediaPlayerPrivateGStreamer* mediaPlayerPrivate, GRefPtr<GstQuery>& query) const
```
ASCIILiteral

calvaris commented on Source/WebCore/platform/gstreamer/GStreamerQuirks.h:
```diff
@@ -62,6 +75,13 @@ class GStreamerQuirk : public GStreamerQuirkBase {
+
+    virtual bool needsBufferingPercentageCorrection() const { return false; }
+    // Returns name of the queried GstElement, or nullptr if no element was queried.
+    virtual const char* queryBufferingPercentage(MediaPlayerPrivateGStreamer*, GRefPtr<GstQuery>&) const { return nullptr; }
```
ASCIILiteral everywhere! /me inserts Oprah meme

calvaris commented on Source/WebCore/platform/gstreamer/GStreamerQuirks.h:
```diff
@@ -107,6 +127,13 @@ class GStreamerQuirksManager : public RefCounted<GStreamerQuirksManager> {
 
+    bool needsBufferingPercentageCorrection() const;
+    // Returns name of the queried GstElement, or nullptr if no element was queried.
+    const char* queryBufferingPercentage(MediaPlayerPrivateGStreamer*, GRefPtr<GstQuery>&) const;
```
oprah

eocanha replied:
Poor Oprah... Don't tell her that in the end, [when she opens the ASCIILiteral wrapped gift](https://github.com/WebKit/WebKit/pull/29998/commits/11a45e5d030157ee664fe02ea9f770aad8dbd41f#diff-4376b854280205494dc9de0f08f985afd4d6584a7487c0a136adac2954c003cdR1340) there's going to be a `const char*` inside. So much wrapping paper wasted...

Should I also use ASCIILiteral in [MediaPlayerPrivateGStreamer::queryBufferingPercentage()](https://github.com/WebKit/WebKit/pull/29998/commits/11a45e5d030157ee664fe02ea9f770aad8dbd41f#diff-4376b854280205494dc9de0f08f985afd4d6584a7487c0a136adac2954c003cdR1336) too? Yeah, sure, let's do it. In the end, I'll still have to force a `static_cast<const char*>()` conversion when I need to [pass it to GST_TRACE_OBJECT()](https://github.com/WebKit/WebKit/pull/29998/commits/11a45e5d030157ee664fe02ea9f770aad8dbd41f#diff-4376b854280205494dc9de0f08f985afd4d6584a7487c0a136adac2954c003cdR1369), but who cares? Memory footprint and impedance caused by wrappers are cheap. 🤷

philn replied:
no cast needed, there is a `characters()` method.

eocanha replied:
Uh... Thank you! 🙏🙏🙏
I somehow missed it in the .h and stubbornly tried to force the usage of `constexpr operator const char*() const` through a static_cast. Though it worked, it felt a bit uncomfortable.

calvaris reviewed (commented):
There are no deal breakers for me. Definitely looking much better.

calvaris commented on Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.h:
```diff
@@ -295,6 +295,36 @@ inline std::optional<T> gstStructureGet(const GstStructure* structure, ASCIILite
 }
 
+template<typename T>
+inline Vector<T> gstStructureGetArray(const GstStructure* structure, ASCIILiteral key)
```
This needs rebasing and moving it to the cpp file.

eocanha replied:
...and to be aware that templates implemented in the cpp file must also be specialized in the cpp file to avoid linking errors, like it was done for [the other template methods](https://github.com/WebKit/WebKit/blob/main/Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp#L1051) moved to the cpp file. 🙄

calvaris commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -1313,6 +1331,56 @@ void MediaPlayerPrivateGStreamer::commitLoad()
+    }
+
+    if (!isQueryOk) {
+        isQueryOk = (m_audioSink && gst_element_query(m_audioSink.get(), query.get()));
```
() are not needed here and below

calvaris commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -2212,44 +2282,75 @@ void MediaPlayerPrivateGStreamer::updateMaxTimeLoaded(double percentage)
         m_fillTimer.stop();
+    else {
+        bool shouldDownload = !m_isLiveStream.value_or(false) && m_preload == MediaPlayer::Preload::Auto && !isMediaDiskCacheDisabled();
+        if (shouldDownload)
```
You can collapse these three lines.

eocanha replied:
Ok, but then I'm mentioning the "should download" fact in a comment. I think it's important for legibility.

calvaris commented on Source/WebCore/platform/gstreamer/GStreamerQuirkBroadcomBase.cpp:
```diff
@@ -0,0 +1,194 @@
+        state.m_streamBufferingLevelMovingAverage.reset(100);
+    int averagedBufferingPercentage = state.m_streamBufferingLevelMovingAverage.accumulate(correctedBufferingPercentage2);
+
+#ifndef GST_DISABLE_GST_DEBUG
```
The ifdef is not preventing a hard and avoidable. computation, I think you could remove it.

calvaris commented on Source/WebCore/platform/gstreamer/GStreamerQuirkBroadcomBase.cpp:
```diff
@@ -0,0 +1,194 @@
+                GRefPtr<GstElement> peerElement = adoptGRef(GST_ELEMENT(gst_pad_get_parent(peerSrcPad.get())));
+
+                // If it's NOT a multiqueue, it's probably a parser like aacparse. We try to traverse before it.
+                if (peerElement && !!g_strcmp0(G_OBJECT_TYPE_NAME(element.get()), "GstMultiQueue")) {
```
Do you really need the `!!`?

calvaris commented on Source/WebCore/platform/gstreamer/GStreamerQuirks.h:
```diff
@@ -44,7 +46,18 @@ class GStreamerQuirkBase {
+
+    // Interface of classes supplied to MediaPlayerPrivateGStreamer to store values that the quirks will need for their job.
+    class GStreamerQuirkState {
+        WTF_MAKE_FAST_ALLOCATED;
```
Maybe making it non-copyable?

eocanha replied:
Wow! This change unveiled a lot of hidden copies that were happening under the hood. I had to define a move constructor ( `= default`, but anyway) both in this base class and in the GStreamerQuirkBroadcomBase subclass, as well as declare some `auto&` instead of `auto` in a couple of places.

I also found out that `makeUnique()` requires a copy, so I had to construct the `std::unique_ptr`s by doing `std::unique_ptr<GStreamerQuirkBase::GStreamerQuirkState>(new GStreamerQuirkBase::GStreamerQuirkState(WTFMove(state)))` in `setQuirkState()` instead. C++ doesn't lose an opportunity to shoot coders in the foot, haha. 🤭

eocanha replied:
The checks that I added (in a debug branch) after adding the move constructor unveiled an even worse problem: [object slicing](https://en.wikipedia.org/wiki/Object_slicing) 😱
On superclass/subclass hierarchies, it's a really bad idea to move (or to copy) an instance of the subclass casted as the superclass, since the specific data of the subclass is **lost** during the move/copy. To prevent this to myself and to future developers, I've declared GStreamerQuirkState as non-copyable (as you said) and also non-movable.

As you suggested to me in private, the right solution is to modify `setQuirkState()` to take pre-made `unique_ptr`, which would be created by the caller (the subclass code, in this case) and would hold an instance of the right class. That instance would never be moved/copied and would therefore not suffer from slicing. And if it did (by some developer mistake in the future), the `WTF_MAKE_NONCOPYABLE` and `WTF_MAKE_NONMOVABLE` macros (which actually `= delete` the copy/move constructors).

philn commented on Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
```diff
@@ -1082,6 +1082,38 @@ StringView gstStructureGetName(const GstStructure* structure)
 }
 
+template<typename T>
+Vector<T> gstStructureGetArray(const GstStructure* structure, ASCIILiteral key)
```
Would be nice to have unit tests for this, see Tools/TestWebKitAPI/Tests/WebCore/gstreamer/GStreamerTest.cpp

eocanha replied:
I'll create a seperate bug for that. This task has been ongoing for 3 months already and I wouldn't like it to delay more its landing now, if possible.

eocanha replied:
Follow up bug created for this unit test writing task: https://bugs.webkit.org/show_bug.cgi?id=280190

eocanha committed:
[GStreamer] Buffering hysteresis and buffering improvements for Broadcom
https://bugs.webkit.org/show_bug.cgi?id=275683

Reviewed by Xabier Rodriguez-Calvar.

The usage of a fixed buffering level to evaluate the need for buffering
results in a too unstable behaviour, with quick changes between states
that may create stuttering.

This commit adds hysteresis to the buffering level, with low and high
watermark levels that trigger "buffering needed" (when below the low
watermark), "buffering completed" (when above high watermark) and "same
state as before" when the level is between both marks.

In addition to that, the pipeline is automatically paused on a seek when
working on stream mode (using GstQueue2 for partial in-memory buffering
instead of using GstDownload for full on-disk download of the media).
This is because, when seeking, rebuffering must occur from scratch to
get the data for the new target position, and buffering level starts
from 0% in that case, which means that buffering is needed and the
pipeline must be paused. Both the m_isBuffering flag and
m_bufferingPercentage are immediately forced to true and 0%, respectively.
The previous values (see explanation below) are also set to the same
values by using the new resetHistory parameter in updateBufferingStatus().
This makes sure that the "end of buffering" condition can be properly
detected.

The buffering status and buffering level computation has been
centralized in updateBufferingStatus(). Both the status and the level
now store the last value, so it's possible to check if there's been any
difference on them and to decide here on the hysteresis status
evolution.

updateStates() is where the consequences of buffering changes happen.
This method now relies in the status and level updates explained in the
previous paragraph. In the specific case of an async change, the current
values for m_isBuffering and m_bufferingPercentage are ignored (previous
ones are used). This is to be able to detect significant changes in the
hysteresis (watermark level crossing) after the async change has
completed. Otherwise, we would be detecting those crossings in a moment
where nothing can be done (pipeline state changes are ignored can can't
be enforced during async state changes).

Some extra improvements have been added to adapt to the special
circumstances of Broadcom devices, where the PlayPump Nexus component
can store a lot of data which is hidden to the vanilla buffering
algorithm. Now that data is had into account to compute the realistic
buffering level. Also, more agile low/high watermarks of 20% and 80% are
used for buffering in the case of playbin2 in stream buffering mode on
Broadcom.

20-80% levels for low/high watermark are now used on all devices
(Broadcom and not Broadcom), since usage tests proved that this makes
the buffering more agile also on Raspberry Pi, for instance.

All these changes are compatible with the Quirks framework. A new
needsBufferingPercentageCorrection() method has been added to enable the
extra buffering logic (which corrects the queue2 buffering level with the
buffer levels of PlayPump and multiqueue) on the platforms that use Nexus.
All the code to implement that logic has been hidden in the
GStreamerQuirkBroadcomBase class, from which both GStreamerQuirkBcmNexus
and GStreamerQuirkBroadcom inherit.

That code needs some elements that in principle would be very coupled with
MediaPlayerPrivateGStreamer, such as vidfilter, queue2, multiqueue and a
MovingAverage to smooth buffering levels. To avoid that coupling, they
are stored in an opaque GStreamerQuirkState, stored by the player private
but implemented by a GStreamerQuirkBroadcomBaseState subclass that only
GStreamerQuirkBroadcomBase knows how to interpret (by casting after
ensuring that the casting is legal).

There's also the problem that multiple competing quirks can be installing
at the same time, but all this buffering percentage correction code must
only be run by one (only one!) of them. To allow the theoretical coexistence
of multiple quirk state, each quirk instance has one associated state, and
the player private maintains that correspondence using a HashMap indexed by
quirk. To ensure that only one quirk is run at each time, the family of methods
in GStreamerQuirksManager only forward the call to the first configured
GStreamerQuirk subclass that declares the need to apply corrections. Any other
configured subclass is ignored.

Finally, some drive-by improvements are present in this commit. For instance,
converting some const char* to ASCIILiteral here and there.

* Source/WebCore/platform/SourcesGStreamer.txt: Added GStreamerQuirkBroadcomBase.
* Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.h:
(WebCore::gstStructureGetArray): Added.
* Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
(WebCore::gstStructureGetArray): Method to get a GstArray GstValue stored into a GstStructure. Returns a Vector of the specified type, which includes basic types and "const GstStructure*".
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::play): Make code compatible with InitiallyPaused.
(WebCore::MediaPlayerPrivateGStreamer::doSeek): Pause pipeline and reset buffering level to 0% in case we're operating in stream mode.
(WebCore::MediaPlayerPrivateGStreamer::queryBufferingPercentage): Get the buffering percentage from the most relevant element and delegate to the quirks.
(WebCore::MediaPlayerPrivateGStreamer::fillTimerFired): Clarify that this method is only used in on-disk buffering. Rely on queryBufferingPercentage() now, instead of manually getting the value from here.
(WebCore::MediaPlayerPrivateGStreamer::handleMessage): Delegate to the quirks to store/delete references to specific elements in the quirks state.
(WebCore::MediaPlayerPrivateGStreamer::processBufferingStats): Delegate to the quierks to apply buffering percentage corrections on platforms that need them.
(WebCore::MediaPlayerPrivateGStreamer::updateBufferingStatus): Update the values of m_wasBuffering, m_isBuffering, m_previousBufferingPercentage, m_bufferingPercentage and m_didDownloadFinish. The first 4 attributes should *ONLY* be modified from this method, in order to guarantee coherency. The only *EXCEPTION* to this rule is the delay which can happen in updateStates(). The new resetHistory parameter sets the value both in the previous value family of attributes and in the current value family of variables. Also, restart the fillTimer if download has restarted (it was never restarted before this patch!)
(WebCore::MediaPlayerPrivateGStreamer::updateStates): Assume that m_wasBuffering, etc. have been precomputed by updateBufferingStatus() before updateStates() is called. In the case of an async state change, delay m_isBuffering and m_bufferingPercentage "to the past" (m_wasBuffering and m_previousBufferingPercentage), so that any change on them can be detected on the next iteration.
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.h: Added new m_wasBuffering and m_previousBufferingPercentage attributes to store previous values of m_isBuffering and m_bufferingPercentage and ease detection of changes. Added quirks state.
(WebCore::MediaPlayerPrivateGStreamer::shouldDownload): Needed by quirks.
(WebCore::MediaPlayerPrivateGStreamer::setQuirkState): Configure a GStreamerQuirkState for a specific GStreamerQuirk on the player private (each quirk has its own state, and in theory there can be many quirk-state stored).
(WebCore::MediaPlayerPrivateGStreamer::quirkState): Get the quirks state associated to a specific quirk from the player private, or nullptr if there's no configured state for that quirk.
* Source/WebCore/platform/gstreamer/GStreamerHolePunchQuirkBcmNexus.h: identifier() is now const ASCIILiteral.
* Source/WebCore/platform/gstreamer/GStreamerHolePunchQuirkFake.h: Ditto.
* Source/WebCore/platform/gstreamer/GStreamerHolePunchQuirkRialto.h: Ditto.
* Source/WebCore/platform/gstreamer/GStreamerHolePunchQuirkWesteros.h: Ditto.
* Source/WebCore/platform/gstreamer/GStreamerQuirkAmLogic.h: Ditto.
* Source/WebCore/platform/gstreamer/GStreamerQuirkBcmNexus.h: Ditto. Inherit from base class.
* Source/WebCore/platform/gstreamer/GStreamerQuirkBroadcom.h: Ditto. Inherit from base class.
* Source/WebCore/platform/gstreamer/GStreamerQuirkBroadcomBase.cpp: Added. Base class for GStreamerQuirkBcmNexus and GStreamerQuirkBroadcom.
(WebCore::GStreamerQuirkBroadcomBase::GStreamerQuirkBroadcomBase): Constructor.
(WebCore::GStreamerQuirkBroadcomBase::queryBufferingPercentage const): Get the buffering percentage from Queue2 if present and the player shouldn't download (streaming mode). Returns the name of the element finally used to query the percentage to, or nullptr if none was queried.
(WebCore::GStreamerQuirkBroadcomBase::correctBufferingPercentage const): On systems using PlayPump (eg: Broadcom Nexus), correct the buffering percentage with the data stored in multiqueue and PlayPump in order to make the buffering percentage more realistic. Uses the MovingAverage. Queries vidfilter and audfilter if needed, so it also works in audio or video-only pipelines.
(WebCore::GStreamerQuirkBroadcomBase::resetBufferingPercentage const): Sets the percentage to a given value in the MovingAverage.
(WebCore::GStreamerQuirkBroadcomBase::setupBufferingPercentageCorrection const): Gets/deletes references to vidfilter, multiqueue and queue2 in the quirks state.
(WebCore::GStreamerQuirkBroadcomBase::ensureState const): Retrieves GStreamerQuirkState associated to this quirk, creating one and storing it on MediaPlayerPrivateGStreamer if not already present.
* Source/WebCore/platform/gstreamer/GStreamerQuirkBroadcomBase.h: Added.
(WebCore::GStreamerQuirkBroadcomBase::needsBufferingPercentageCorrection const): True for this class and derived ones.
(WebCore::GStreamerQuirkBroadcomBase::MovingAverage::MovingAverage): Initialize an array of "length" number of elements, asserting if length is too big that the values would create overflow when computing the average.
(WebCore::GStreamerQuirkBroadcomBase::MovingAverage::reset): Reset all the values to a given value (usually zero).
(WebCore::GStreamerQuirkBroadcomBase::MovingAverage::accumulate): Insert a new value in the array and recompute the moving average.
* Source/WebCore/platform/gstreamer/GStreamerQuirkRealtek.h: identifier() is now const ASCIILiteral.
* Source/WebCore/platform/gstreamer/GStreamerQuirkRialto.h: Ditto.
* Source/WebCore/platform/gstreamer/GStreamerQuirkWesteros.h: Ditto.
* Source/WebCore/platform/gstreamer/GStreamerQuirks.cpp: Added new methods.
(WebCore::GStreamerQuirksManager::GStreamerQuirksManager): Adapt to ASCIILiteral quirk identifier.
(WebCore::GStreamerQuirksManager::createAudioSink): Ditto.
(WebCore::GStreamerQuirksManager::createWebAudioSink): Ditto.
(WebCore::GStreamerQuirksManager::createHolePunchVideoSink): Ditto.
(WebCore::GStreamerQuirksManager::isHardwareAccelerated const): Ditto.
(WebCore::GStreamerQuirksManager::audioVideoDecoderFactoryListType const): Ditto.
(WebCore::GStreamerQuirksManager::getAdditionalPlaybinFlags const): Ditto.
(WebCore::GStreamerQuirksManager::needsBufferingPercentageCorrection const): True if any configured quirk needs it. The first configured quirk that needs it will be the one responsible to fulfill all the requests for this family of methods. The other ones will be ignored, even if they also need percentage correction. This is to avoid more than one implementation doing the corrections, as we don't want them to accumulate and over-correct.
(WebCore::GStreamerQuirksManager::queryBufferingPercentage const): Forward query request to the first configured quirk needing correction.
(WebCore::GStreamerQuirksManager::correctBufferingPercentage const): Forward correction request to the first configured quirk needing correction.
(WebCore::GStreamerQuirksManager::resetBufferingPercentage const): Ditto, but for reset.
(WebCore::GStreamerQuirksManager::setupBufferingPercentageCorrection const): Ditto, but for setup.
* Source/WebCore/platform/gstreamer/GStreamerQuirks.h:
(WebCore::GStreamerQuirkBase::GStreamerQuirkState::GStreamerQuirkState): Constructor.
(WebCore::GStreamerQuirk::needsBufferingPercentageCorrection const): Default empty implementation.
(WebCore::GStreamerQuirk::queryBufferingPercentage const): Ditto.
(WebCore::GStreamerQuirk::correctBufferingPercentage const): Ditto.
(WebCore::GStreamerQuirk::resetBufferingPercentage const): Ditto.
(WebCore::GStreamerQuirk::setupBufferingPercentageCorrection const): Ditto.
* LayoutTests/platform/glib/TestExpectations: Reported new failing test as https://bugs.webkit.org/show_bug.cgi?id=280069, but IMHO it's not a regression. The test was passing before by sheer luck.

Canonical link: https://commits.webkit.org/284072@main
